bson/decoder.go

```diff
@@ -31,8 +31,10 @@ var decPool = sync.Pool{
 // A Decoder reads and decodes BSON documents from a stream. It reads from a bsonrw.ValueReader as
 // the source of BSON data.
 type Decoder struct {
-	dc bsoncodec.DecodeContext
-	vr bsonrw.ValueReader
+	dc               bsoncodec.DecodeContext
+	vr               bsonrw.ValueReader
+	defaultDocumentM bool
```

Comment:
We do this instead of setting directly on `dc` because the `Decoder` exposes a setter for the `DecodeContext`:  `(*Decoder).SetContext`.  If the user set the default document behavior and _then_ the context, the default behavior will be overwritten.  Persisting the behavior on the `Decoder` allows us to set the `documentType` on the decoder before the `DecodeValue` call.

```go
dec.DefaultDocumentD()

// If we didn't do it this way, this would cause the default behavior to be overwritten.
dec.SetContext(bsoncodec.DecodeContext{})
```

Reply:
Optional: Consider leaving a comment about this choice in the code as well so future devs understand the decision.